runtime/Java/src/org/antlr/v4/runtime/ParserRuleContext.java

```diff
@@ -101,9 +102,8 @@ public void copyFrom(ParserRuleContext ctx) {
 			this.children = new ArrayList<>();
 			// reset parent pointer for any error nodes
 			for (ParseTree child : ctx.children) {
-				if ( child instanceof ErrorNodeImpl ) {
-					this.children.add(child);
-					((ErrorNodeImpl) child).parent = this;
+				if ( child instanceof ErrorNode ) {
+					addChild((ErrorNode)child);
```

Comment:
Well, that is implicit knowledge, which is not visible at this point in code and for some targets like C++ it is very important to know what is managing the memory. And if you one day decide to keep the source context around you certainly forgot by that time that it contains nodes which are supposed to be in a different context.